### PR TITLE
feat: add slot event part docs to foundation

### LINKED ISSUE #5904: feat: add slot, event and css shadow part JSDoc tags to Fast-Foundation components
Update comments in Fast-Foundation components to include all necessary @slot, @fires, and @csspart information. 

The CEM Analyzer is not capable of parsing the templates in order to automatically determine slots and CSS Shadow Parts but can read them out of the JSDOC tag comments placed in the comment block above the class declaration in the [component].ts files. Unfortunately, this means that this information must be added to the comments manually in order to for them to be rendered into the README.md files that then become the documentation pages on the fast web site.

See https://custom-elements-manifest.open-wc.org/analyzer/getting-started/#documenting-your-components for documentation on what JSDoc tags are supported by the CEM Analyzer and their syntax.

For example:
/fast-foundation/src/text-field/text-field.ts

    /**
     * A Text Field Custom HTML Element.
     * Based largely on the {@link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/text | <input type="text" /> element }.
     *
     * @slot start - Content before the input control. Useful for placing an icon in front of the input.
     * @slot end - Content after the input control. Useful for placing an icon after the input.
     * @fires change - Fires a custom 'change' event when the input content changes.
     * @csspart label - Part for the label before the input.
     * @csspart root - Part for the root container of the input.
     * @csspart control - Part for the input control.
     *
     * @public
     */
    export class TextField extends FormAssociatedTextField {

This is how the information is rendered in the component's README.md file:

#### Events

| Name     | Type | Description                                                   | Inherited From |
| -------- | ---- | ------------------------------------------------------------- | -------------- |
| `change` |      | Fires a custom 'change' event when the input content changes. |                |

#### CSS Parts

| Name      | Description                               |
| --------- | ----------------------------------------- |
| `label`   | Part for the label before the input.      |
| `root`    | Part for the root container of the input. |
| `control` | Part for the input control.               |

#### Slots

| Name    | Description                                                                         |
| ------- | ----------------------------------------------------------------------------------- |
| `start` | Content before the input control. Useful for placing an icon in front of the input. |
| `end`   | Content after the input control. Useful for placing an icon after the input.        |

Closes: #5863

### PR DESCRIPTION
**Pull request by chrisdholt**
# Pull Request

## 📖 Description
This PR does an initial pass to add jsdocs to enable support for slots, events, and parts in the CEM analyzer.

Two components have some gaps due to how composable they are - Calendar and Data Grid.

### 🎫 Issues
closes #5904

### General

- [ ] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [ ] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [ ] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

File: packages/web-components/fast-foundation/src/accordion-item/accordion-item.ts [hunk 1]
```diff
@@ -22,6 +22,22 @@ export type AccordionItemOptions = FoundationElementDefinition &
 
 /**
  * An individual item in an {@link @microsoft/fast-foundation#(Accordion:class) }.
+ *
+ * @slot start - Content which can be provided between the heading and the icon
+ * @slot end - Content which can be provided between the start slot and icon
+ * @slot heading - Content which serves as the accordion item heading and text of the expand button
+ * @slot - The default slot for accordion item content
+ * @slot - expanded-icon - The expanded icon
```
Comment (williamw2) [inline packages/web-components/fast-foundation/src/accordion-item/accordion-item.ts:30]: The analyzer uses the hyphen as the separator between the name and the description. These hyphens between @slot and the name should be removed.

**Review by williamw2 — APPROVED** [review]

File: packages/web-components/fast-foundation/src/accordion-item/accordion-item.ts (hunk 1 above)
Comment (chrisdholt) [reply packages/web-components/fast-foundation/src/accordion-item/accordion-item.ts:30]: Yeah, I did almost all of these manually, I'll do one last check for any time it's not the default slot.

**Review by EisenbergEffect — APPROVED** [review]

**Comment by scomea** [comment]
> @scomea are you able to add this context for data-grid once this PR goes in? I'm did not feel familiar enough with the code to address everything.

Ok.
https://github.com/microsoft/fast/issues/5913

**Comment by nicholasrice** [comment]
Wow, thanks Chris! I'm so happy we're finally able to get these flags in. One thing that catches my eye here is that it documents the foundation components as if they are a single component, template *and* class, where in this package those two are never actually combined. For example, the ES6 class conceptually never has css parts or slots because those are template concerns. I don't know if we should change anything now because we would likely need to adjust the CEM code, but to me it would make more sense if the class and templates are both documented w/ their respective TSDoc flags. Any thoughts on that? We can take to an issue if needed.

**Comment by chrisdholt** [comment]
> Wow, thanks Chris! I'm so happy we're finally able to get these flags in. One thing that catches my eye here is that it documents the foundation components as if they are a single component, template _and_ class, where in this package those two are never actually combined. For example, the ES6 class conceptually never has css parts or slots because those are template concerns. I don't know if we should change anything now because we would likely need to adjust the CEM code, but to me it would make more sense if the class and templates are both documented w/ their respective TSDoc flags. Any thoughts on that? We can take to an issue if needed.

I fired this off without much investigation after a few conversations with @williamw2, so I have zero thoughts. Totally open, just going by the initial convention shared. I’ll default to him :)

**Comment by williamw2** [comment]
@nicholasrice I don't think that's going to work. The problem is in how we declare the templates as a const function that returns a template. The CEM analyzer sees it as just a function and seems to ignore any jsdoc tags that don't apply to functions. Then the markdown generator simply renders it in a "function" table with just a name and description. I'm currently manually excluding the template functions from the markdown output as they didn't seem useful. The plan is to eventually replace the CEM markdown generation with a custom solution. Maybe we can try to address this then.

**Pull request merged**